File: app/src/main/java/com/kira/kdownloader/engine/YtDlpVersionPolicy.java
package com.kira.kdownloader.engine;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class YtDlpVersionPolicy {
    private static final Pattern DATE_PATTERN = Pattern.compile("(\\d{4})\\.(\\d{1,2})\\.(\\d{1,2})");

    private YtDlpVersionPolicy() {}

    static boolean shouldInstallBundled(String installedVersion, String bundledVersion) {
        if (bundledVersion.equals(installedVersion)) return false;
        Integer installedDate = installedVersion == null ? null : dateKey(installedVersion);
        if (installedDate == null) return true;
        Integer bundledDate = dateKey(bundledVersion);
        return bundledDate == null ? !installedVersion.equals(bundledVersion) : installedDate < bundledDate;
    }

    private static Integer dateKey(String version) {
        Matcher match = DATE_PATTERN.matcher(version);
        if (!match.find()) return null;
        try {
            return Integer.parseInt(match.group(1)) * 10_000
                    + Integer.parseInt(match.group(2)) * 100
                    + Integer.parseInt(match.group(3));
        } catch (NumberFormatException ignored) {
            return null;
        }
    }
}

File: app/src/main/java/com/kira/kdownloader/engine/YtDlpVersionPolicy.kt
package com.kira.kdownloader.engine

internal object YtDlpVersionPolicy {
    private val datePattern = Regex("(\\d{4})\\.(\\d{1,2})\\.(\\d{1,2})")

    /**
     * A stale or unreadable installed copy must be replaced. A newer self-updated copy is kept.
     */
    fun shouldInstallBundled(installedVersion: String?, bundledVersion: String): Boolean {
        if (installedVersion == bundledVersion) return false

        val installedDate = installedVersion?.let(::dateKey) ?: return true
        val bundledDate = dateKey(bundledVersion) ?: return installedVersion != bundledVersion
        return installedDate < bundledDate
    }

    private fun dateKey(version: String): Int? {
        val match = datePattern.find(version) ?: return null
        val (year, month, day) = match.destructured
        return year.toIntOrNull()?.times(10_000)
            ?.plus((month.toIntOrNull() ?: return null) * 100)
            ?.plus(day.toIntOrNull() ?: return null)
    }
}
